Authentication

UJKhokhar opened the pull request:
- [x] Create tokens for users
- [x] Authenticate user actions

chikamichi commented:
Nice! The basic mechanism is there.

Two possible improvements:

* `authenticate_or_request_with_http_token` respond with HTML. For consistency, we'd like to return JSON. I know the resource you used as a reference has something on this ;)
* When you generate the token for a user, it's unlikely but not entirely impossible you create a duplicate token. It'd be better to check for duplicates, and keep generating tokens until you get a unique one. In order to do this, you'd need something like the `while` construct, and to fetch (`User.find(token: …)`) users to check whether a token is already in use.

UJKhokhar commented:
Thanks for the notes. I added authentication checks to ApiController and also created a custom json response. 

I tried implementing a while test for tokens here but couldn't get it to work: https://gist.github.com/UJKhokhar/d28a636dec14d173a1ce

I then ran a migration, added ```auth_token``` to users, indexed and also added a unique option. 

However, after running ```rake db:reset``` I'm receiving an error: ```Could not find a valid mapping for #<User id: 1...```

chikamichi commented:
Excellent. I answered directly on the gist.

chikamichi commented:
By the way, here's something that may interest you: https://github.com/gonzalo-bulnes/simple_token_authentication

chikamichi commented on app/controllers/api/api_controller.rb:
```diff
@@ -1,3 +1,21 @@
+
+  def render_unauthorized
+    self.headers['WWW-Authenticate'] = 'Token realm="Application"'
+    render json: 'Bad credentials', status: 401
```
A bit late, but I've just noticed that we're not sending valid JSON back here. It should be a hash or an array. Maybe something like this:

``` ruby
render json: '{"error" => "Bad credentials"}', status: 401
```

You can verify that `'Bad credentials'` is not valid on [jsonlint.com](http://jsonlint.com/).
